File: go/fixture/go-mini/internal/repository/mem_store.go
package repository

import (
	"context"
	"fmt"
	"sync"

	"example.com/go-mini/internal/models"
)

// MemStore keeps the fleet in memory. It is the store behind `svc --dry-run`,
// where an operator replays recorded heartbeats without writing anything to
// disk.
type MemStore struct {
	mu      sync.Mutex
	devices map[string]models.Device
}

// NewMemStore creates a new MemStore.
func NewMemStore() *MemStore {
	return &MemStore{devices: map[string]models.Device{}}
}

// Get gets a device.
func (m *MemStore) Get(ctx context.Context, tenant, id string) (models.Device, error) {
	if err := ctx.Err(); err != nil {
		return models.Device{}, fmt.Errorf("mem store: get: %w", err)
	}
	m.mu.Lock()
	defer m.mu.Unlock()
	d, ok := m.devices[key(tenant, id)]
	if !ok {
		return models.Device{}, ErrNotFound
	}
	return d, nil
}

// Save saves a device.
func (m *MemStore) Save(ctx context.Context, d models.Device) error {
	if err := ctx.Err(); err != nil {
		return fmt.Errorf("mem store: save: %w", err)
	}
	m.mu.Lock()
	defer m.mu.Unlock()
	m.devices[key(d.Tenant, d.ID)] = d
	return nil
}

// List lists all devices.
func (m *MemStore) List(ctx context.Context) ([]models.Device, error) {
	if err := ctx.Err(); err != nil {
		return nil, fmt.Errorf("mem store: list: %w", err)
	}
	m.mu.Lock()
	defer m.mu.Unlock()
	return sorted(m.devices), nil
}

File: py/fixture/py-mini/internal/repository/mem_store.py
"""An in-memory device store."""

import threading

from internal.models.device import Device
from internal.repository.store import NotFoundError, _key


class MemStore:
    """Keeps the fleet in memory.

    It is the store behind `svc --dry-run`, where an operator replays
    recorded heartbeats without writing anything to disk.
    """

    def __init__(self) -> None:
        """Create a new MemStore."""
        self._guard = threading.Lock()
        self._devices: dict[str, Device] = {}

    def get(self, tenant: str, device_id: str) -> Device:
        """Get a device."""
        with self._guard:
            d = self._devices.get(_key(tenant, device_id))
        if d is None:
            raise NotFoundError(f"device {tenant}/{device_id} not found")
        return d

    def save(self, d: Device) -> None:
        """Save a device."""
        with self._guard:
            self._devices[_key(d.tenant, d.id)] = d

    def list_devices(self) -> list[Device]:
        """List all devices."""
        with self._guard:
            devices = dict(self._devices)
        if not devices:
            return []
        return [devices[k] for k in sorted(devices)]
